elasticapm/contrib/asyncio/traces.py

```diff
@@ -0,0 +1,69 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import functools
+
+from elasticapm.traces import capture_span, error_logger, execution_context
+from elasticapm.utils import get_name_from_func
+
+
+class async_capture_span(capture_span):
+    def __call__(self, func):
+        self.name = self.name or get_name_from_func(func)
+
+        @functools.wraps(func)
+        async def decorated(*args, **kwds):
+            async with self:
+                return await func(*args, **kwds)
+
+        return decorated
+
+    async def __aenter__(self):
+        transaction = execution_context.get_transaction()
+        if transaction and transaction.is_sampled:
+            return transaction.begin_span(
+                self.name,
+                self.type,
+                context=self.extra,
+                leaf=self.leaf,
+                labels=self.labels,
+                span_subtype=self.subtype,
+                span_action=self.action,
+                sync=False,
+                start=self.start,
+            )
```

Comment:
Just an idea: this+capture_span could be made a bit more DRY by one of these approaches:
 - add a sync slot to capture_span, and setting it to False in the async_capture_span constructor; then `__aenter__`/`__aexit__` could delegate to `__enter__`/`__exit__` as they don't do anything synchronous
 - split the code in `__enter__`/`__exit__` into separate methods that take a `sync` arg, and calling them with `sync=False` from `__aenter__`/`__aexit__`

Reply:
I'm a bit wary of introducing more layers into this code path, as every layer ends up being captured as span stack frame. We could refactor this a bit to capture the stack trace directly in the context manager and pass it into the `begin_span` methods.

If it's ok with you, I'll leave this as is, and do the refactor in a follow-up PR. I'd like to merge this soonish, as I currently have to do a lot of git dances locally to merge this into other branches that depend on it, only to back it out again before pushing.

Reply:
> I'm a bit wary of introducing more layers into this code path, as every layer ends up being captured as span stack frame. We could refactor this a bit to capture the stack trace directly in the context manager and pass it into the begin_span methods.

Couldn't you just add 1 to skip_frames when calling `end`?

> If it's ok with you, I'll leave this as is, and do the refactor in a follow-up PR. I'd like to merge this soonish, as I currently have to do a lot of git dances locally to merge this into other branches that depend on it, only to back it out again before pushing.

Sounds good. Up to you anyway, I just thought it might help keep maintenance costs down.